k8s/upgrade/src/bin/upgrade-job/upgrade/utils.rs

```diff
@@ -209,3 +210,126 @@ pub(crate) async fn data_plane_is_upgraded(
 
     Ok(true)
 }
+
+/// Cordon storage node.
+pub(crate) async fn cordon_storage_node(
+    node_id: &str,
+    cordon_label: &str,
+    rest_client: &RestClientSet,
+) -> Result<()> {
+    let cordon_label = cordon_label.to_string();
+    loop {
+        let storage_node =
+            rest_client
+                .nodes_api()
+                .get_node(node_id)
+                .await
+                .context(GetStorageNode {
+                    node_id: node_id.to_string(),
+                })?;
+
+        match storage_node
+            .into_body()
+            .spec
+            .ok_or(
+                EmptyStorageNodeSpec {
+                    node_id: node_id.to_string(),
+                }
+                .build(),
+            )?
+            .cordondrainstate
+        {
+            Some(CordonDrainState::cordonedstate(cordon_state))
+                if cordon_state.cordonlabels.contains(&cordon_label) =>
+            {
+                info!(node.id = %node_id, "Cordoned {PRODUCT} Node");
+                return Ok(());
+            }
+            _ => {
+                rest_client
+                    .nodes_api()
+                    .put_node_cordon(node_id, cordon_label.as_str())
+                    .await
+                    .context(CordonStorageNode {
+                        node_id: node_id.to_string(),
+                    })?;
+
+                info!(node.id = %node_id, "Put cordon label for {PRODUCT} Node");
```

Comment:
We have to return Ok(()) here?

Reply:
No. We check again in the next iteration of the loop, and return Ok(()) if we find the cordon label set. Is the cordon label put guaranteed to succeed? I am using it this way, because of how drain works, it is possible that drain might be draining and reach drained at a later time.

Reply:
But why, if put succeeds then it's done, what else are you waiting for?

Reply:
Alright. But the returns stay right?

Reply:
Removed the loops.